File: leetcode/0032.go
package leetcode

func longestValidParentheses(s string) int {
	ans, stack := 0, make([]int, 0)
	stack = append(stack, -1)
	for i, ch := range s {
		if ch == '(' {
			stack = append(stack, i)
		} else {
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				stack = append(stack, i)
			} else {
				ans = max(ans, i-stack[len(stack)-1])
			}
		}
	}
	return ans
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: leetcode/0032.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int ans = 0;
        stack<int> ss;
        ss.emplace(-1);
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                ss.emplace(i);
            } else {
                ss.pop();
                if (ss.empty()) {
                    ss.emplace(i);
                } else {
                    ans = max(ans, i - ss.top());
                }
            }
        }
        return ans;
    }
};